presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -791,7 +803,31 @@ public void addUtilizedTableColumnReferences(AccessControlInfo accessControlInfo
 
     public Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> getTableColumnReferencesForAccessControl(Session session)
     {
-        Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> references = isCheckAccessControlOnUtilizedColumnsOnly(session) ? utilizedTableColumnReferences : tableColumnReferences;
+        Map<AccessControlInfo, Map<QualifiedObjectName, Set<String>>> references;
+        if (!isCheckAccessControlWithSubFieldInfo(session)) {
+            references = isCheckAccessControlOnUtilizedColumnsOnly(session) ? utilizedTableColumnReferences : tableColumnReferences;
+        }
+        else if (!isCheckAccessControlOnUtilizedColumnsOnly(session)) {
+            references = tableColumnSubFieldReferences;
+        }
+        else {
+            // TODO: Properly support utilized column check. Currently, we prune whole columns, if they are not utilized.
+            // We need to generalize it and exclude unutilized subfield references independently.
+            references = tableColumnSubFieldReferences.entrySet().stream()
+                    .collect(toImmutableMap(
+                            Map.Entry::getKey, accessControlEntry ->
+                                    accessControlEntry.getValue().entrySet().stream().collect(toImmutableMap(
+                                            Map.Entry::getKey, tableEntry -> tableEntry.getValue().stream().filter(
+                                                    column -> {
+                                                        Map<QualifiedObjectName, Set<String>> utilizedTableReferences = utilizedTableColumnReferences.get(accessControlEntry.getKey());
+                                                        if (utilizedTableReferences == null) {
+                                                            return false;
+                                                        }
+                                                        Set<String> utilizedColumns = utilizedTableReferences.get(tableEntry.getKey());
+                                                        return utilizedColumns != null && utilizedColumns.contains(column.split("\\.")[0]);
```

Comment:
Let's try to leverage `Subfield` class as much as possible. For example, can it be used in some interfaces? Also, I guess `SubfieldTokenizer` can be a helper to structure a subfield